File: app/src/main/java/com/kabirnayeem99/paymentpaid/data/db/WorkDatabase.java
package com.kabirnayeem99.paymentpaid.data.db;

import android.content.Context;

import androidx.room.Database;
import androidx.room.Room;
import androidx.room.RoomDatabase;

import com.kabirnayeem99.paymentpaid.data.db.entities.Work;

@Database(entities = {Work.class}, version = 1, exportSchema = false)
public abstract class WorkDatabase extends RoomDatabase {
    public static final String DB_NAME = "payment_paid_db";
    public static WorkDatabase instance;


    public static WorkDatabase getInstance(Context context) {
        if (instance == null) {
            instance = Room.databaseBuilder(context.getApplicationContext(),
                    WorkDatabase.class, WorkDatabase.DB_NAME).fallbackToDestructiveMigration()
                    .build();
        }
        return instance;
    }

    public abstract WorkDao workDao();
}

File: app/src/main/java/com/kabirnayeem99/paymentpaid/data/db/WorkDatabase.kt
package com.kabirnayeem99.paymentpaid.data.db

import android.content.Context
import androidx.room.Database
import androidx.room.Room
import androidx.room.RoomDatabase
import com.kabirnayeem99.paymentpaid.data.db.entities.Work
import com.kabirnayeem99.paymentpaid.utils.Constants

@Database(entities = [Work::class], version = 1, exportSchema = false)
abstract class WorkDatabase : RoomDatabase() {

    abstract fun getWorkDao(): WorkDao

    companion object {

        /*
        the singleton pattern is a software design pattern
        that restricts the instantiation of a class to one "single" instance.
         */

        @Volatile
        private var instance: WorkDatabase? = null

        private var LOCK = Any()

        operator fun invoke(context: Context) = instance ?: synchronized(LOCK) {
            instance ?: createDatabase(context).also { instance = it }
        }

        private fun createDatabase(context: Context) =
                Room.databaseBuilder(
                        context.applicationContext,
                        WorkDatabase::class.java,
                        Constants.DB_NAME
                ).build()

    }
}